orc8r/cloud/go/services/ctraced/obsidian/handlers/handlers.go

```diff
@@ -0,0 +1,246 @@
+		return obsidian.HttpError(err, http.StatusInternalServerError)
+	}
+
+	ret := map[string]*models.CallTrace{}
+	for _, ctEnt := range callTraces {
+		ret[ctEnt.Key] = (&models.CallTrace{}).FromEntity(ctEnt)
+	}
+	return c.JSON(http.StatusOK, ret)
+}
+
+func GetCreateCallTraceHandlerFunc(client GWCtracedClient) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		networkID, nerr := obsidian.GetNetworkId(c)
+		if nerr != nil {
+			return nerr
+		}
+		cfg := &models.CallTraceConfig{}
+		if err := c.Bind(cfg); err != nil {
+			return obsidian.HttpError(err, http.StatusBadRequest)
+		}
+		ctr := &models.CallTrace{
+			Config: cfg,
+			State: &models.CallTraceState{
+				CallTraceAvailable: false,
+				CallTraceEnding:    false,
+			},
+		}
+		if err := ctr.ValidateModel(); err != nil {
+			return obsidian.HttpError(err, http.StatusBadRequest)
+		}
+
+		req, err := buildStartTraceRequest(cfg)
+		if err != nil {
+			return obsidian.HttpError(errors.Wrap(err, "failed to build call trace request"), http.StatusInternalServerError)
+		}
+
+		resp, err := client.StartCallTrace(networkID, cfg.GatewayID, req)
+		if err != nil {
+			return obsidian.HttpError(errors.Wrap(err, "failed to start call trace"), http.StatusInternalServerError)
+		}
+		if resp.Success == false {
+			return obsidian.HttpError(errors.New("failed to start call trace"), http.StatusInternalServerError)
+		}
+
+		createdEntity := ctr.ToEntity()
+		_, err = configurator.CreateEntity(networkID, createdEntity, serdes.Entity)
+		if err != nil {
+			return obsidian.HttpError(errors.Wrap(err, "failed to create call trace"), http.StatusInternalServerError)
+		}
+		return c.JSON(http.StatusCreated, string(cfg.TraceID))
+	}
+}
+
+func GetCallTrace(c echo.Context) error {
+	callTrace, err := getCallTrace(c)
+	if err != nil {
+		return err
+	}
+	return c.JSON(http.StatusOK, callTrace)
+}
+
+func GetUpdateCallTraceHandlerFunc(client GWCtracedClient, storage storage.CtracedStorage) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		networkID, callTraceID, nerr := getNetworkIDAndCallTraceID(c)
+		if nerr != nil {
+			return nerr
+		}
+		mutableCallTrace := &models.MutableCallTrace{}
+		if err := c.Bind(mutableCallTrace); err != nil {
+			return obsidian.HttpError(err, http.StatusBadRequest)
+		}
+		if err := mutableCallTrace.ValidateModel(); err != nil {
+			return obsidian.HttpError(err, http.StatusBadRequest)
+		}
+
+		callTrace, err := getCallTrace(c)
+		if err != nil {
+			return err
+		}
+
+		req := &protos.EndTraceRequest{}
+		resp, err := client.EndCallTrace(networkID, callTrace.Config.GatewayID, req)
+		if err != nil {
+			return err
+		}
+		if resp.Success == false {
+			return obsidian.HttpError(errors.New("Failed to end call trace"), http.StatusInternalServerError)
+		}
+
+		err = storage.StoreCallTrace(networkID, callTraceID, resp.TraceContent)
+		if err != nil {
+			return obsidian.HttpError(errors.Wrap(err, "failed to save call trace data"), http.StatusInternalServerError)
```

Comment:
add more information in the error message, include networkID/gatewayID/callTraceID perhaps.